### PR TITLE
[SPARK-34383][SS] Optimize WAL commit phase via reducing cost of filesystem operations

### PR DESCRIPTION
**Pull request by HeartSaVioR**
### What changes were proposed in this pull request?

This PR proposes to optimize WAL commit phase via following changes:

* cache offset log to avoid FS get operation per batch
* just directly delete instead of employing FS list operation on purge

### Why are the changes needed?

There're inefficiency on WAL commit phase which can be easily optimized via using a small driver memory.

1. To provide the offset metadata to source side (via `source.commit()`), we read offset metadata for previous batch from file system, which is probably written by this driver in previous batches. Caching it into driver memory would reduce the get operation.
2. Spark calls purge against offset log & commit log per batch, which calls list operation. If the previous batch succeeded to purge, the current batch just needs to check one batch which can be simply done via direct delete operation, instead of calling list operation.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Manually tested with additional debug log. (Verified that cache is used, cache keeps the size as 2, only one delete call is used instead of list call)

Did some experiment with simple rate to console query. (NOTE: wasn't done with master branch - tested against Spark 2.4.x, but WAL commit phase hasn't been changed AFAIK during these versions)

AWS S3 + S3 guard:

> before the patch

aws-before

> after the patch

aws-after

Azure:

> before the patch

azure-before

> after the patch

azure-after

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala
Comment (HeartSaVioR) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:57]: Note for reviewers: see MicroBatchExecution / ContinuousExecution around calling `source.commit()`. To provide the offset metadata to source side we read offset metadata for previous batch, which is probably written by this driver in previous batches.

Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:57]: `TreeMap` seems overkill to me if we only cache at most 2 batches. How about using a 2-size array? We can put/get the elements by the index of (batchId % 2).

For example:

```scala
private val cachedMetadata = Array[OffsetSeq](null, null)

// put
cachedMetadata(batchId % 2) = metadata

// get
Option(cachedMetadata.get(batchId % 2)).orElse(super.get(batchId))
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala
Comment (Ngone51) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:249]: Usually, `possibleTargetBatchIds.length` should be 1?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala
Comment (Ngone51) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:253]: This confused me. Why not blindly delete the path as the original way does?

**Comment by Ngone51** [comment]
I wonder what if we purge the offset/commit logs in batches? For example, we can purge logs every `minLogEntriesToMaintain` batches. This probably reduces more operations on filesystem.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:249]: Yes usually the value will be 1, once purge is executed once via listing. I just don't set it to 1 because I don't want to make the logic be tight to MicroBatchExecution/ContinuousExecution. (Someone outside of Spark may leverage this as well.)

That's a magic number based on heuristic so I don't mind too much about the value (just should be greater than 1), but considering the cost of list vs exist, the ideal threshold wouldn't be just 1.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:253]: Original way is not blindly deleting. It does list to enumerate the paths and deletes them. Right?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:57]: TreeMap approach is pretty much intuitive compared to this, and the overhead should be pretty much trivial compared to FS ops. Let's not do over-engineering.

**Comment by HeartSaVioR** [comment]
> I wonder what if we purge the offset/commit logs in batches? For example, we can purge logs every minLogEntriesToMaintain batches. This probably reduces more operations on filesystem.

I also considered this (even with asynchronous manner) but didn't address it altogether. While this PR doesn't change anything on end users perspective, the proposal changes the remaining number of log files. Probably better to make it separate and hear voices on making change.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:57]: Or I should just do this in MicroBatchExecution / ContinuousExecution instead - then the code doesn't need to be generalized. Let me have a time to revisit this.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:57]: Hmm... it looks easier to change here and reflect to both micro-batch and continuous mode. Let's leave this as it is.

I would worry about not coupling metadata with batch ID - it could bring critical issue on committing to source when something is slightly messed up. e.g. In file stream source we delete source files based on the commit. We cannot trade-off between correctness and performance. I'd prefer safer approach unless it's identified to have a noticeable performance impact.

**Comment by HeartSaVioR** [comment]
cc. @tdas @zsxwing @jose-torres @viirya @gaborgsomogyi @xuanyuanking

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala
Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:57]: > e.g. In file stream source we delete source files based on the commit. 

I don't understand why this could bring a correctness issue (I'm still very new to streaming so I may miss some use cases). The only problem I can imagine now is `get(batchId)` may lead to the correctness issue as `batchId` like 2 or 4 results in the same index. But this can also be addressed by saving the `batchId` at the same, e.g., `Array[(batchId, OffsetSeq)]`, which I think also addresses your concern for batchId coupling.

BTW, I did a minor performance test between these two ways, and seems it does have a performance impact:

```scala

// ~270ms
timeTakeMs {
  treeMap.put(1, "1")
  treeMap.put(2, "3")
  treeMap.put(3, "3")
  treeMap.headMap(2, true).clear()
}

// ~0ms
timeTakeMs {
  array(0) = "0"
  array(1) = "1"
  array(0) = "2"
}
```

Thoughts?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala
Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:253]: I see, although seems blindly deleting is also fine since it doesn't throw the exception if the file doesn't exist and reduces one more filesystem operation.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:57]:
```
import java.lang.{Long => JLong}
import java.util.{ArrayList, Collections, TreeMap}

def c(treeMap: TreeMap[Long, String]): Long = {
  val t1 = System.nanoTime()
  treeMap.put(1, "1")
  treeMap.put(2, "3")
  treeMap.put(3, "3")
  treeMap.headMap(2, true).clear()
  (System.nanoTime() - t1)
}

def d(treeMap: TreeMap[Long, String], idx: Long, value: String): Long = {
  val t1 = System.nanoTime()
  treeMap.put(idx, value)
  treeMap.headMap(idx - 2, true).clear()
  (System.nanoTime() - t1)
}

def experimentC(): Unit = {
  val latencies = new ArrayList[JLong]()
  val warmupCount = 1000000
  val runCount = 10000000

  (1 to warmupCount).foreach { _ =>
    val t = new java.util.TreeMap[Long, String]()
    c(t)
  }

  (1 to runCount).foreach { _ =>
    val t = new java.util.TreeMap[Long, String]()
    latencies.add(JLong.valueOf(c(t)))
  }

  java.util.Collections.sort(latencies)

  printLatencies(latencies)
}

def experimentD(): Unit = {
  val latencies = new ArrayList[JLong]()
  val warmupCount = 1000000
  val runCount = 10000000

  val t = new java.util.TreeMap[Long, String]()
  (1 to warmupCount).foreach { idx =>
    d(t, idx, idx.toString)
  }

  val t2 = new java.util.TreeMap[Long, String]()
  (1 to runCount).foreach { idx =>
    latencies.add(JLong.valueOf(d(t2, idx, idx.toString)))
  }

  printLatencies(latencies)
}

def printLatencies(latencies: ArrayList[JLong]): Unit = {
  val arraySize = latencies.size()
  val minIdx = 0
  val maxIdx = arraySize - 1
  val percentile50 = (arraySize * 0.5).toInt
  val percentile90 = (arraySize * 0.9).toInt
  val percentile99 = (arraySize * 0.99).toInt
  val percentile999 = (arraySize * 0.999).toInt
  val percentile9999 = (arraySize * 0.9999).toInt
  val percentile99999 = (arraySize * 0.99999).toInt
  val percentile999999 = (arraySize * 0.999999).toInt

  java.util.Collections.sort(latencies)

  Seq(minIdx, percentile50, percentile90, percentile99, percentile999, percentile9999, percentile99999, percentile999999, maxIdx).foreach { idx =>
    printLatency(latencies, idx)
  }  
}

def printLatency(latencies: ArrayList[JLong], idx: Int): Unit = {
  println(s"$idx th : ${latencies.get(idx) / 1000} microseconds = ${latencies.get(idx) / 1000000} milliseconds")
}

// experimentC()

/*
0 th : 0 microseconds = 0 milliseconds
5000000 th : 0 microseconds = 0 milliseconds
9000000 th : 0 microseconds = 0 milliseconds
9900000 th : 0 microseconds = 0 milliseconds
9990000 th : 1 microseconds = 0 milliseconds
9999000 th : 9 microseconds = 0 milliseconds
9999900 th : 37 microseconds = 0 milliseconds
9999990 th : 223 microseconds = 0 milliseconds
9999999 th : 53612 microseconds = 53 milliseconds
*/

experimentD()

/*
0 th : 0 microseconds = 0 milliseconds
5000000 th : 0 microseconds = 0 milliseconds
9000000 th : 0 microseconds = 0 milliseconds
9900000 th : 0 microseconds = 0 milliseconds
9990000 th : 0 microseconds = 0 milliseconds
9999000 th : 6 microseconds = 0 milliseconds
9999900 th : 25 microseconds = 0 milliseconds
9999990 th : 150 microseconds = 0 milliseconds
9999999 th : 57887 microseconds = 57 milliseconds
*/
```

2018 13-inch MBP, i7 quad-core 2.7Ghz

```
./bin/spark-shell --driver-memory 2g
...
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 3.0.1
      /_/

Using Scala version 2.12.10 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_191)
```

Still think this really matters?

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:57]: Even without warmup (commenting out),

```
// experimentC()

...
9999990 th : 1632 microseconds = 1 milliseconds
9999999 th : 60999 microseconds = 60 milliseconds

// experimentD()

...
9999990 th : 321 microseconds = 0 milliseconds
9999999 th : 35074 microseconds = 35 milliseconds
```

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:253]: Yeah that also makes sense. I'm not sure about how much the cost would be saved though. Let me play with this a bit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala
Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:57]: Ok..I saw where the problem is with my test. You're right the latency is trivial.

I'm not against your solution here. But since we've reached here, I'd like to mention one more thing that TreeMap tends to produce the instant object `AscendingSubMap` for each batch while Array doesn't. Although, It might also be trivial.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala
Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:253]: Sure, please. I'm fine either way unless there's a noticeable perf difference.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:57]: This is another sort of micro-optimization; realistic latency of micro-batch is 1s+ (doesn't matter if we consider very tight micro-batch, like 500ms) and we worry about creating "an" object per such period which will be marked as "unused" after couple of batches.

This is the clear example why micro-optimization is bad without understanding full context - optimization should evaluate about the impact and proceed only when it contributes at least 1% (I'd rather not even concern about 1% though if the code is more intuitive).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:253]: (Just wanted to mention; the case what I'm considering is when the file doesn't exist - then the case would be changed to exist vs delete. I'm going to evaluate this because I don't know the cost comparison between exist and delete on non-exist - if the cost difference is significant and exist is faster, it's going to be some sort of probability/heuristic. If not, we should simply try calling delete.)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala
Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:57]: Thanks for bringing the details, it makes sense to me. But please also note that I wouldn't raise such optimization changes if the `TreeMap` is already an existing implementation. However, for a PR, I think it's good to have more inputs regardless of the final decision.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:57]: The reason I am concerning on the input on possible micro-optimization is that it changes the viewpoint of the codebase to performance and doesn't evaluate based on other viewpoints. There're bunch of points to discuss on this change, is this better to be put to OffsetSeq or better to place it on both MicroBatchExecution/ContinuousExecution (and why), which code would be most intuitive to maintainers or contributors, etc. Once the performance perspective comes in and gains traction, other perspectives tend to be disregarded. That should be just one of viewpoints, and shouldn't be overvalued.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:253]: I thought about this a bit more, and I admit I was wrong about the assumption; log file may not be existed. It could be for some batches after restarting from checkpoint, but once the query runs more than the threshold (100 batches by default), the log file "should" exist.

Thanks for raising this. Will make a change and run simple test before pushing.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala
Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:57]: I know your concern. I like those general optimization ideas too instead of hacks(micro-optimization). Here the argument may be that you think the `Array` way is not intuitive enough. Well, I'm actually neutral on this as it's also not too complex as it is. Let's move on since I already +1 for the current solution :)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:253]: Ahh... there's a roadblock on the details - FileSystemBasedCheckpointFileManager.delete logs INFO message when the file doesn't exist. FileContextBasedCheckpointFileManager.delete doesn't log any message though.

Probably OK to remove INFO message on FileSystemBasedCheckpointFileManager.delete as FileSystemBasedCheckpointFileManager is used as a fail-back of FileContextBasedCheckpointFileManager and FileContextBasedCheckpointFileManager covers the major usage.

Comment (steveloughran) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:253]: delete() is a no-op if the file doesn't exist; it always has to do the check to see if its there (and whether a file/dir), so wrapping it with another check is superfluous. Azure abfs will suffer here, HDFS less, but still two RPCs with one block for NN read access, one for NN r/w. I don't know anything about HDFS NN locking so can't comment there

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala:253]: Looks like adding the check doesn't contribute noticeable latency;

before the change

```
grep "walCommit took" driver_stderr.log | awk '{print $7}' | head -n 1100 | tail -n 1000 | datamash max 1 min 1 mean 1 median 1 perc:90 1 perc:95 1 perc:99 1
1874	152	242.131	220.5	306	342	654.84
```

after the change

```
grep "walCommit took" driver_stderr.log | awk '{print $7}' | head -n 1100 | tail -n 1000 | datamash max 1 min 1 mean 1 median 1 perc:90 1 perc:95 1 perc:99 1
1882	156	252.162	231.5	311	357.2	559.16
```

but I agree the check is unnecessary. Will remove it.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: Addressed via 03801f7

Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: thanks for the investigation

**Comment by Ngone51** [comment]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala
Comment (Ngone51) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:50]: What about moving `cachedMetadata` to  `HDFSMetadataLog`? Seems like other metadata logs can benefit too.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:50]: There're different usage patterns and characteristics against these metadata logs, and the change is rather only bound to the usage pattern of offsetLog.
e.g. commitLog is only added per batch, and read only when query is started or restored from checkpoint.

Comment (Ngone51) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:50]: @zsxwing thoughts?

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:50]: You can check by yourself: find `commitLog.get(` and you get nothing except tests. find `commitLog.getLatest(` and realize that is only happening when starting/restarting query. Simple enough?

**Comment by HeartSaVioR** [comment]
Kindly reminder @tdas @zsxwing @jose-torres @viirya @gaborgsomogyi @xuanyuanking

**Comment by HeartSaVioR** [comment]
retest this, please

**Comment by HeartSaVioR** [comment]
retest this, please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CheckpointFileManager.scala
Comment (xuanyuanking) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CheckpointFileManager.scala:280]: Just to clarify, this log was deleted per https://github.com/apache/spark/pull/31495#discussion_r571870795

**Review by xuanyuanking — APPROVED** [review]
LGTM

**Comment by HeartSaVioR** [comment]
Unfortunately this still lacks of committers' reviews though it got two +1 from active contributors.
Kindly reminder @tdas @zsxwing @jose-torres @viirya and possibly another eye of review @gaborgsomogyi

**Comment by HeartSaVioR** [comment]
Friendly reminder,  @tdas @zsxwing @jose-torres @viirya @gaborgsomogyi

**Comment by gaborgsomogyi** [comment]
I've picked this up but it takes some time to test it properly...

**Comment by gaborgsomogyi** [comment]
I've double checked the performance on Azure only and seen similar trending.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala
Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/OffsetSeqLog.scala:57]: I've gone through this thread and changing the current solution would introduce quite some non-trivial code structures and the gain would be not significant from memory point of view so +1 to keep it as-is.

**Comment by gaborgsomogyi** [comment]
`the proposal changes the remaining number of log files (based on the interval on purge).`
+1 on this, changes which would impact users such a way is not something what would be a good idea (at least not in a performance tuning PR). I'm on the same side not to make such sudden move.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CheckpointFileManager.scala
Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CheckpointFileManager.scala:280]: @xuanyuanking @HeartSaVioR not sure I get the full point here. We've removed the log which notified the user that the file was not able to remove and now we have a new log like this:
`logTrace(s"Removed metadata log file: $path")`
The plan is that if the user sees this log then it was successful and when nothing is in the log then it has failed?
Can you enlighten me please?

**Comment by gaborgsomogyi** [comment]
All in all I have 1 [comment ](https://github.com/apache/spark/pull/31495#discussion_r595205302)and when this answered and/or resolved I'm happy to take care of this PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CheckpointFileManager.scala
Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CheckpointFileManager.scala:280]: I'm quoting my comment:

> FileSystemBasedCheckpointFileManager.delete logs INFO message when the file doesn't exist. FileContextBasedCheckpointFileManager.delete doesn't log any message though.
>
> Probably OK to remove INFO message on FileSystemBasedCheckpointFileManager.delete as FileSystemBasedCheckpointFileManager is used as a fail-back of FileContextBasedCheckpointFileManager (especially Spark logs warn message when failing back) and FileContextBasedCheckpointFileManager covers the major usage.

We decided to "blindly" call `delete()` without checking existence in prior, as file system will do it instead and we can save one op call. Leaving this log would provide false-alarm, as it's "expected" even if there's no file, so we have to delete the log line.

Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CheckpointFileManager.scala:280]: In the meantime I was digging the code and now I see why the if was needed. Calling `delete` blindly is really a speedup.
False-alarm is the key to my question, thanks!

**Review by gaborgsomogyi — APPROVED** [review]
LGTM.

**Comment by gaborgsomogyi** [comment]
@tdas @zsxwing @jose-torres @viirya I'm leaving this open until beginning of next week.
If no further comments/suggestions are coming then I'm intended to merge it.

**Review by viirya — APPROVED** [review]

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CheckpointFileManager.scala
Comment (steveloughran) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CheckpointFileManager.scala:280]: > Calling delete blindly is really a speedup.

mmm. Those HEAD requests add up, don't they?

FWIW, Hadoop-3.3.1 adds a lot of IOStatistics collection on s3a and abfs operations, for which there's an interface implemented by, amongst other things: input streams, output streams, filesystems, remote iterators. But toString() logs this stuff too. So if you log an FS.toString() at debug you will see these. We are doing more than just counting #of calls, we are collecting min/max/mean performance. LIST is the killer -especially if you have versioning turned on on S3 and then do lots of overwrite/delete of objects.

Comment (gaborgsomogyi) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CheckpointFileManager.scala:280]: You're right, HEAD requests add up and I've formed my statement roughly.
According to my measurements the nominal performance better even with blind `delete`.
To tell the truth not measured independently how much caching brings and how much blind `delete` wastes.
If you have a good idea to make it better it's more than welcome.

Comment (HeartSaVioR) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CheckpointFileManager.scala:280]: I've already experimented and noticed both will have no noticeable difference.
https://github.com/apache/spark/pull/31495#discussion_r572473687

But that's completely based on implementation details; implementations can leverage the benefit of doing less operations if external storage could simply do what the operation associated with delete promises the same contract delete() in Hadoop filesystem promises.

**Comment by HeartSaVioR** [comment]
Thanks @gaborgsomogyi for merging and everyone for reviewing!

(A general guide to @gaborgsomogyi - it'd be nice for merger to comment which branches he/she merged and pushed the changes. JIRA issue would have such information, but in case of information mismatch.)

**Comment by gaborgsomogyi** [comment]
Good point, Is it a feature of `merge_spark_pr.py`?

**Comment by gaborgsomogyi** [comment]
Additionally I've just set up the jira part because some python packages were missing to handle it automatically.

**Comment by HeartSaVioR** [comment]
> @HeartSaVioR Good point, is it a feature of merge_spark_pr.py?

Nope, we manually did it, with saying to contributor "thanks for the contribution". So that's something like formal process :)

**Comment by gaborgsomogyi** [comment]
Ah, ok. Need to steal the customs so I'm on it :)

**Comment by gaborgsomogyi** [comment]
Merged to master.

**Pull request closed without merging**
